In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from datetime import datetime
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots


In [ ]:
# Set plotting style
plt.style.use('fivethirtyeight')
sns.set_palette("deep")

Upload your dataset


In [ ]:
#from google.colab import files

# Upload the file from your local machine
pbj_df = pd.read_csv(r"/content/PBJ_Daily_Nurse_Staffing_Q2_2024.csv", encoding='latin1')
# Replace with the name of your file
pbj_df



<ipython-input-13-40aa500f5347>:4: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pbj_df = pd.read_csv(r"/content/PBJ_Daily_Nurse_Staffing_Q2_2024.csv", encoding='latin1')


PROVNUM                       PROVNAME          CITY STATE  \
0         15009       BURNS NURSING HOME, INC.  RUSSELLVILLE    AL   
1         15009       BURNS NURSING HOME, INC.  RUSSELLVILLE    AL   
2         15009       BURNS NURSING HOME, INC.  RUSSELLVILLE    AL   
3         15009       BURNS NURSING HOME, INC.  RUSSELLVILLE    AL   
4         15009       BURNS NURSING HOME, INC.  RUSSELLVILLE    AL   
...         ...                            ...           ...   ...   
1325319  745040  THE SARAH ROBERTS FRENCH HOME   SAN ANTONIO    TX   
1325320  745040  THE SARAH ROBERTS FRENCH HOME   SAN ANTONIO    TX   
1325321  745040  THE SARAH ROBERTS FRENCH HOME   SAN ANTONIO    TX   
1325322  745040  THE SARAH ROBERTS FRENCH HOME   SAN ANTONIO    TX   
1325323  745040  THE SARAH ROBERTS FRENCH HOME   SAN ANTONIO    TX   

        COUNTY_NAME  COUNTY_FIPS  CY_Qtr  WorkDate  MDScensus  Hrs_RNDON  ...  \
0          Franklin           59  2024Q2  20240401         51      10.77  ...   
1          Franklin           59  2024Q2  20240402         52       8.43  ...   
2          Franklin           59  2024Q2  20240403         53      11.13  ...   
3          Franklin           59  2024Q2  20240404         52      12.27  ...   
4          Franklin           59  2024Q2  20240405         52       4.95  ...   
...             ...          ...     ...       ...        ...        ...  ...   
1325319       Bexar           29  2024Q2  20240626         44       8.00  ...   
1325320       Bexar           29  2024Q2  20240627         44       8.00  ...   
1325321       Bexar           29  2024Q2  20240628         45       8.00  ...   
1325322       Bexar           29  2024Q2  20240629         45       0.00  ...   
1325323       Bexar           29  2024Q2  20240630         45       0.00  ...   

         Hrs_LPN_ctr  Hrs_CNA  Hrs_CNA_emp  Hrs_CNA_ctr  Hrs_NAtrn  \
0                0.0   160.08       160.08          0.0        0.0   
1                0.0   135.95       135.95          0.0        0.0   
2                0.0   150.31       150.31          0.0        0.0   
3                0.0   133.01       133.01          0.0        0.0   
4                0.0   137.92       137.92          0.0        0.0   
...              ...      ...          ...          ...        ...   
1325319          0.0    93.99        93.99          0.0        0.0   
1325320          0.0    85.99        85.99          0.0        0.0   
1325321          0.0    74.05        74.05          0.0        0.0   
1325322          0.0    55.44        55.44          0.0        0.0   
1325323          0.0    66.81        66.81          0.0        0.0   

         Hrs_NAtrn_emp  Hrs_NAtrn_ctr  Hrs_MedAide  Hrs_MedAide_emp  \
0                  0.0            0.0         0.00             0.00   
1                  0.0            0.0         0.00             0.00   
2                  0.0            0.0         0.00             0.00   
3                  0.0            0.0         0.00             0.00   
4                  0.0            0.0         0.00             0.00   
...                ...            ...          ...              ...   
1325319            0.0            0.0        15.60            15.60   
1325320            0.0            0.0        13.70            13.70   
1325321            0.0            0.0        17.18            17.18   
1325322            0.0            0.0        20.92            20.92   
1325323            0.0            0.0        13.58            13.58   

         Hrs_MedAide_ctr  
0                    0.0  
1                    0.0  
2                    0.0  
3                    0.0  
4                    0.0  
...                  ...  
1325319              0.0  
1325320              0.0  
1325321              0.0  
1325322              0.0  
1325323              0.0  

[1325324 rows x 33 columns]

Inspect Missing Values


In [ ]:
provider_df = pd.read_csv(r"/content/NH_ProviderInfo_Feb2025.csv", encoding='latin1')
# Replace with the name of your file
provider_df

CMS Certification Number (CCN)  \
0                             015009   
1                             015010   
2                             015012   
3                             015014   
4                             015015   
...                              ...   
14777                         745049   
14778                         745050   
14779                         745051   
14780                         745052   
14781                         745054   

                                           Provider Name  \
0                               BURNS NURSING HOME, INC.   
1                         COOSA VALLEY HEALTHCARE CENTER   
2                             HIGHLANDS HEALTH AND REHAB   
3            EASTVIEW REHABILITATION & HEALTHCARE CENTER   
4                          PLANTATION MANOR NURSING HOME   
...                                                  ...   
14777  LAS ALTURAS NURSING & TRANSITIONAL CARE BROWNS...   
14778        RIVERSIDE NURSING AND REHABILITATION CENTER   
14779  FIVE POINTS NURSING & REHABILITATION OF COLLEG...   
14780                   HCA HOUSTON HEALTHCARE SOUTHEAST   
14781                       LEGACY ESTATE LONG TERM CARE   

                          Provider Address        City/Town State  ZIP Code  \
0                     701 MONROE STREET NW     RUSSELLVILLE    AL     35653   
1                   260 WEST WALNUT STREET        SYLACAUGA    AL     35150   
2                      380 WOODS COVE ROAD       SCOTTSBORO    AL     35768   
3                 7755 FOURTH AVENUE SOUTH       BIRMINGHAM    AL     35206   
4              6450 OLD TUSCALOOSA HIGHWAY         MC CALLA    AL     35111   
...                                    ...              ...   ...       ...   
14777                  180 EAST PRICE ROAD      BROWNSVILLE    TX     78521   
14778                     1555 BANDERA HWY        KERRVILLE    TX     78028   
14779                   3105 CORSAIR DRIVE  COLLEGE STATION    TX     77845   
14780  4801 EAST SAM HOUSTON PARKWAY SOUTH         PASADENA    TX     77505   
14781                       10133 HWY 16 N         COMANCHE    TX     76442   

       Telephone Number  Provider SSA County Code County/Parish  \
0            2563324110                       290      Franklin   
1            2562495604                       600     Talladega   
2            2562183708                       350       Jackson   
3            2058330146                       360     Jefferson   
4            2054776161                       360     Jefferson   
...                 ...                       ...           ...   
14777        9563032002                       240       Cameron   
14778        5129920058                       734          Kerr   
14779        8178328016                       190        Brazos   
14780        7133592000                       610        Harris   
14781        2548794970                       321      Comanche   

                               Ownership Type  ...  \
0                    For profit - Corporation  ...   
1                    For profit - Corporation  ...   
2                         Government - County  ...   
3                     For profit - Individual  ...   
4                     For profit - Individual  ...   
...                                       ...  ...   
14777                For profit - Corporation  ...   
14778  For profit - Limited Liability company  ...   
14779                For profit - Corporation  ...   
14780                For profit - Corporation  ...   
14781          Government - Hospital district  ...   

       Number of Citations from Infection Control Inspections  \
0                                                    NaN        
1                                                    0.0        
2                                                    NaN        
3                                                    0.0        
4                                                    NaN        
...        

In [ ]:
pbj_df.info()
provider_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1325324 entries, 0 to 1325323
Data columns (total 33 columns):
 #   Column            Non-Null Count    Dtype  
---  ------            --------------    -----  
 0   PROVNUM           1325324 non-null  object 
 1   PROVNAME          1325324 non-null  object 
 2   CITY              1325324 non-null  object 
 3   STATE             1325324 non-null  object 
 4   COUNTY_NAME       1325324 non-null  object 
 5   COUNTY_FIPS       1325324 non-null  int64  
 6   CY_Qtr            1325324 non-null  object 
 7   WorkDate          1325324 non-null  int64  
 8   MDScensus         1325324 non-null  int64  
 9   Hrs_RNDON         1325324 non-null  float64
 10  Hrs_RNDON_emp     1325324 non-null  float64
 11  Hrs_RNDON_ctr     1325324 non-null  float64
 12  Hrs_RNadmin       1325324 non-null  float64
 13  Hrs_RNadmin_emp   1325324 non-null  float64
 14  Hrs_RNadmin_ctr   1325324 non-null  float64
 15  Hrs_RN            1325324 non-null  float64
 16  

In [ ]:
pbj_df.head()
provider_df.head()

CMS Certification Number (CCN)                                Provider Name  \
0                         015009                     BURNS NURSING HOME, INC.   
1                         015010               COOSA VALLEY HEALTHCARE CENTER   
2                         015012                   HIGHLANDS HEALTH AND REHAB   
3                         015014  EASTVIEW REHABILITATION & HEALTHCARE CENTER   
4                         015015                PLANTATION MANOR NURSING HOME   

              Provider Address     City/Town State  ZIP Code  \
0         701 MONROE STREET NW  RUSSELLVILLE    AL     35653   
1       260 WEST WALNUT STREET     SYLACAUGA    AL     35150   
2          380 WOODS COVE ROAD    SCOTTSBORO    AL     35768   
3     7755 FOURTH AVENUE SOUTH    BIRMINGHAM    AL     35206   
4  6450 OLD TUSCALOOSA HIGHWAY      MC CALLA    AL     35111   

   Telephone Number  Provider SSA County Code County/Parish  \
0        2563324110                       290      Franklin   
1        2562495604                       600     Talladega   
2        2562183708                       350       Jackson   
3        2058330146                       360     Jefferson   
4        2054776161                       360     Jefferson   

             Ownership Type  ...  \
0  For profit - Corporation  ...   
1  For profit - Corporation  ...   
2       Government - County  ...   
3   For profit - Individual  ...   
4   For profit - Individual  ...   

   Number of Citations from Infection Control Inspections  Number of Fines  \
0                                                NaN                     1   
1                                                0.0                     0   
2                                                NaN                     0   
3                                                0.0                     0   
4                                                NaN                     0   

   Total Amount of Fines in Dollars Number of Payment Denials  \
0                           23989.0                         0   
1                               0.0                         0   
2                               0.0                         0   
3                               0.0                         0   
4                               0.0                         0   

  Total Number of Penalties                                       Location  \
0                         1     701 MONROE STREET NW,RUSSELLVILLE,AL,35653   
1                         0      260 WEST WALNUT STREET,SYLACAUGA,AL,35150   
2                         0        380 WOODS COVE ROAD,SCOTTSBORO,AL,35768   
3                         0   7755 FOURTH AVENUE SOUTH,BIRMINGHAM,AL,35206   
4                         0  6450 OLD TUSCALOOSA HIGHWAY,MC CALLA,AL,35111   

  Latitude Longitude  Geocoding Footnote Processing Date  
0  34.5149   -87.736                 NaN      2025-02-01  
1  33.1637   -86.254                 NaN      2025-02-01  
2  34.6611   -86.047                 NaN      2025-02-01  
3  33.5595   -86.722                 NaN      2025-02-01  
4  33.3222   -87.034                 NaN      2025-02-01  

[5 rows x 103 columns]

In [ ]:

# Step 3: Clean and prepare the data
print("\nCleaning and preparing data...")

# Rename the CCN column in provider_df to match PROVNUM in pbj_df
provider_df = provider_df.rename(columns={'CMS Certification Number (CCN)': 'PROVNUM'})

# Convert WorkDate to datetime (assuming it's in YYYYMMDD format)
pbj_df['WorkDate'] = pd.to_datetime(pbj_df['WorkDate'].astype(str), format='%Y%m%d')



Cleaning and preparing data...


In [ ]:
# Add week number and year-week columns for time-based analysis
pbj_df['Week'] = pbj_df['WorkDate'].dt.isocalendar().week
pbj_df['YearWeek'] = pbj_df['WorkDate'].dt.year.astype(str) + '-' + pbj_df['WorkDate'].dt.isocalendar().week.astype(str).str.zfill(2)

# Add weekend indicator
pbj_df['IsWeekend'] = pbj_df['WorkDate'].dt.dayofweek >= 5  # 5=Saturday, 6=Sunday

# Calculate total hours for each staff type
print("Calculating total hours by staff type...")
for staff_type in ['RN', 'LPN', 'CNA', 'NAtrn', 'MedAide', 'RNDON', 'RNadmin', 'LPNadmin']:
    # Total hours
    if f'Hrs_{staff_type}' not in pbj_df.columns:
        pbj_df[f'Hrs_{staff_type}'] = pbj_df[f'Hrs_{staff_type}_emp'] + pbj_df[f'Hrs_{staff_type}_ctr']

# Calculate total hours (employee and contractor) across all staff types
pbj_df['Total_Hrs'] = (pbj_df['Hrs_RN'] + pbj_df['Hrs_LPN'] + pbj_df['Hrs_CNA'] +
                       pbj_df['Hrs_NAtrn'] + pbj_df['Hrs_MedAide'] +
                       pbj_df['Hrs_RNDON'] + pbj_df['Hrs_RNadmin'] + pbj_df['Hrs_LPNadmin'])

pbj_df['Total_Hrs_emp'] = (pbj_df['Hrs_RN_emp'] + pbj_df['Hrs_LPN_emp'] + pbj_df['Hrs_CNA_emp'] +
                          pbj_df['Hrs_NAtrn_emp'] + pbj_df['Hrs_MedAide_emp'] +
                          pbj_df['Hrs_RNDON_emp'] + pbj_df['Hrs_RNadmin_emp'] + pbj_df['Hrs_LPNadmin_emp'])

pbj_df['Total_Hrs_ctr'] = (pbj_df['Hrs_RN_ctr'] + pbj_df['Hrs_LPN_ctr'] + pbj_df['Hrs_CNA_ctr'] +
                          pbj_df['Hrs_NAtrn_ctr'] + pbj_df['Hrs_MedAide_ctr'] +
                          pbj_df['Hrs_RNDON_ctr'] + pbj_df['Hrs_RNadmin_ctr'] + pbj_df['Hrs_LPNadmin_ctr'])

# Calculate contractor dependency ratio (CDR)
pbj_df['CDR'] = pbj_df['Total_Hrs_ctr'] / pbj_df['Total_Hrs']
pbj_df['CDR'].fillna(0, inplace=True)  # Replace NaN with 0 where Total_Hrs might be 0

# Calculate Hours Per Patient Day (HPPD) for each staff type
pbj_df['HPPD_RN'] = pbj_df['Hrs_RN'] / pbj_df['MDScensus']
pbj_df['HPPD_LPN'] = pbj_df['Hrs_LPN'] / pbj_df['MDScensus']
pbj_df['HPPD_CNA'] = pbj_df['Hrs_CNA'] / pbj_df['MDScensus']
pbj_df['HPPD_Total'] = pbj_df['Total_Hrs'] / pbj_df['MDScensus']

# Replace infinite values with NaN
pbj_df.replace([np.inf, -np.inf], np.nan, inplace=True)

# Step 4: Merge the datasets
print("Merging datasets...")
merged_df = pbj_df.merge(provider_df, on='PROVNUM', how='left')

# Step 5: Analytics for Recommendation 1 - Target High Contractor Dependency Facilities
print("\nAnalyzing contractor dependency and volatility...")

# Calculate facility-level metrics
facility_metrics = pbj_df.groupby('PROVNUM').agg(
    Total_Hours=('Total_Hrs', 'sum'),
    Employee_Hours=('Total_Hrs_emp', 'sum'),
    Contractor_Hours=('Total_Hrs_ctr', 'sum'),
    Avg_Census=('MDScensus', 'mean'),
    Max_Census=('MDScensus', 'max'),
).reset_index()

# Calculate facility-level CDR
facility_metrics['Facility_CDR'] = facility_metrics['Contractor_Hours'] / facility_metrics['Total_Hours']
facility_metrics['Facility_CDR'].fillna(0, inplace=True)

# Calculate weekly CDR for each facility to measure volatility
weekly_cdr = pbj_df.groupby(['PROVNUM', 'YearWeek']).agg(
    Weekly_Total_Hours=('Total_Hrs', 'sum'),
    Weekly_Contractor_Hours=('Total_Hrs_ctr', 'sum'),
).reset_index()

weekly_cdr['Weekly_CDR'] = weekly_cdr['Weekly_Contractor_Hours'] / weekly_cdr['Weekly_Total_Hours']
weekly_cdr['Weekly_CDR'].fillna(0, inplace=True)

# Calculate CDR volatility (standard deviation of weekly CDR)
cdr_volatility = weekly_cdr.groupby('PROVNUM').agg(
    CDR_Volatility=('Weekly_CDR', 'std'),
    Mean_Weekly_CDR=('Weekly_CDR', 'mean'),
    Weeks_Count=('Weekly_CDR', 'count')
).reset_index()

# Only keep facilities with enough data for volatility calculation (at least 4 weeks)
cdr_volatility = cdr_volatility[cdr_volatility['Weeks_Count'] >= 4]

# Merge volatility data into facility metrics
facility_metrics = facility_metrics.merge(cdr_volatility[['PROVNUM', 'CDR_Volatility', 'Mean_Weekly_CDR']],
                                         on='PROVNUM', how='left')

# Add facility state from the original dataset
facility_state = pbj_df[['PROVNUM', 'STATE']].drop_duplicates()
facility_metrics = facility_metrics.merge(facility_state, on='PROVNUM', how='left')

# Define high-opportunity facilities (high CDR and high volatility)
national_avg_cdr = facility_metrics['Facility_CDR'].mean()
cdr_threshold = 0.35  # 35% contractor dependency
volatility_threshold = facility_metrics['CDR_Volatility'].quantile(0.75)  # Top 25% volatility

facility_metrics['High_Opportunity'] = ((facility_metrics['Facility_CDR'] > cdr_threshold) &
                                       (facility_metrics['CDR_Volatility'] > volatility_threshold))

# Count high-opportunity facilities
high_opp_count = facility_metrics['High_Opportunity'].sum()
total_facilities = len(facility_metrics)
high_opp_percentage = (high_opp_count / total_facilities) * 100

print(f"National average contractor dependency ratio: {national_avg_cdr:.1%}")
print(f"High-opportunity facilities: {high_opp_count} ({high_opp_percentage:.1f}% of total)")

# Analyze geographic distribution of high-opportunity facilities
geo_dist = facility_metrics[facility_metrics['High_Opportunity']].groupby('STATE').size().reset_index(name='Count')
geo_dist['Percentage'] = geo_dist['Count'] / high_opp_count * 100
geo_dist = geo_dist.sort_values('Count', ascending=False).head(10)

# Step 6: Analytics for Recommendation 2 - LPN/CNA Focus Program
print("\nAnalyzing staffing gaps by position type...")

# Calculate staffing levels by position type
staffing_by_type = pbj_df.groupby(['PROVNUM', 'IsWeekend']).agg(
    RN_Hours=('Hrs_RN', 'sum'),
    LPN_Hours=('Hrs_LPN', 'sum'),
    CNA_Hours=('Hrs_CNA', 'sum'),
    RN_Contractor_Hours=('Hrs_RN_ctr', 'sum'),
    LPN_Contractor_Hours=('Hrs_LPN_ctr', 'sum'),
    CNA_Contractor_Hours=('Hrs_CNA_ctr', 'sum'),
    Census=('MDScensus', 'mean')
).reset_index()

# Calculate HPPD (Hours Per Patient Day) by position type
staffing_by_type['RN_HPPD'] = staffing_by_type['RN_Hours'] / staffing_by_type['Census']
staffing_by_type['LPN_HPPD'] = staffing_by_type['LPN_Hours'] / staffing_by_type['Census']
staffing_by_type['CNA_HPPD'] = staffing_by_type['CNA_Hours'] / staffing_by_type['Census']

# Calculate contractor percentage by position type
staffing_by_type['RN_Contractor_Pct'] = staffing_by_type['RN_Contractor_Hours'] / staffing_by_type['RN_Hours']
staffing_by_type['LPN_Contractor_Pct'] = staffing_by_type['LPN_Contractor_Hours'] / staffing_by_type['LPN_Hours']
staffing_by_type['CNA_Contractor_Pct'] = staffing_by_type['CNA_Contractor_Hours'] / staffing_by_type['CNA_Hours']

# Replace NaN and infinity values
staffing_by_type.replace([np.inf, -np.inf], np.nan, inplace=True)
staffing_by_type.fillna(0, inplace=True)

# Calculate weekday vs weekend staffing patterns
weekday_weekend = staffing_by_type.pivot_table(
    index='PROVNUM',
    columns='IsWeekend',
    values=['RN_HPPD', 'LPN_HPPD', 'CNA_HPPD', 'Census',
            'RN_Contractor_Pct', 'LPN_Contractor_Pct', 'CNA_Contractor_Pct']
)

# Create a flattened version of the pivot table
weekday_weekend_flat = pd.DataFrame(index=weekday_weekend.index)
for col in weekday_weekend.columns:
    position, weekend = col
    weekday_weekend_flat[f'{position}_{"Weekend" if weekend else "Weekday"}'] = weekday_weekend[col]

# Calculate weekend/weekday ratios
for position in ['RN_HPPD', 'LPN_HPPD', 'CNA_HPPD', 'Census']:
    weekday_weekend_flat[f'{position}_Ratio'] = (
        weekday_weekend_flat[f'{position}_Weekend'] / weekday_weekend_flat[f'{position}_Weekday']
    )

weekday_weekend_flat.reset_index(inplace=True)
weekday_weekend_flat.replace([np.inf, -np.inf], np.nan, inplace=True)

# Calculate staffing gaps
# Define minimum staffing requirements based on literature/regulations
min_rn_hppd = 0.75  # Minimum RN hours per patient day
min_lpn_hppd = 0.55  # Minimum LPN hours per patient day
min_cna_hppd = 2.0   # Minimum CNA hours per patient day

# Calculate staffing gap ratio (SGR)
weekday_weekend_flat['RN_SGR_Weekday'] = weekday_weekend_flat['RN_HPPD_Weekday'] / min_rn_hppd
weekday_weekend_flat['RN_SGR_Weekend'] = weekday_weekend_flat['RN_HPPD_Weekend'] / min_rn_hppd
weekday_weekend_flat['LPN_SGR_Weekday'] = weekday_weekend_flat['LPN_HPPD_Weekday'] / min_lpn_hppd
weekday_weekend_flat['LPN_SGR_Weekend'] = weekday_weekend_flat['LPN_HPPD_Weekend'] / min_lpn_hppd
weekday_weekend_flat['CNA_SGR_Weekday'] = weekday_weekend_flat['CNA_HPPD_Weekday'] / min_cna_hppd
weekday_weekend_flat['CNA_SGR_Weekend'] = weekday_weekend_flat['CNA_HPPD_Weekend'] / min_cna_hppd

# Identify facilities with critical staffing gaps
weekday_weekend_flat['RN_Critical_Gap'] = weekday_weekend_flat['RN_SGR_Weekend'] < 0.75
weekday_weekend_flat['LPN_Critical_Gap'] = weekday_weekend_flat['LPN_SGR_Weekend'] < 0.75
weekday_weekend_flat['CNA_Critical_Gap'] = weekday_weekend_flat['CNA_SGR_Weekend'] < 0.75

# Identify facilities with low contractor utilization
weekday_weekend_flat['RN_Low_Contractor'] = weekday_weekend_flat['RN_Contractor_Pct_Weekend'] < 0.1
weekday_weekend_flat['LPN_Low_Contractor'] = weekday_weekend_flat['LPN_Contractor_Pct_Weekend'] < 0.1
weekday_weekend_flat['CNA_Low_Contractor'] = weekday_weekend_flat['CNA_Contractor_Pct_Weekend'] < 0.1

# Define high-potential facilities for the LPN/CNA program
weekday_weekend_flat['LPN_High_Potential'] = (
    weekday_weekend_flat['LPN_Critical_Gap'] &
    weekday_weekend_flat['LPN_Low_Contractor'] &
    (weekday_weekend_flat['Census_Weekend'] > 0.9 * weekday_weekend_flat['Census_Weekday'])
)

weekday_weekend_flat['CNA_High_Potential'] = (
    weekday_weekend_flat['CNA_Critical_Gap'] &
    weekday_weekend_flat['CNA_Low_Contractor'] &
    (weekday_weekend_flat['Census_Weekend'] > 0.9 * weekday_weekend_flat['Census_Weekday'])
)

weekday_weekend_flat['LPN_CNA_High_Potential'] = (
    weekday_weekend_flat['LPN_High_Potential'] |
    weekday_weekend_flat['CNA_High_Potential']
)

# Count high-potential facilities
lpn_cna_high_potential_count = weekday_weekend_flat['LPN_CNA_High_Potential'].sum()
print(f"High-potential facilities for LPN/CNA program: {lpn_cna_high_potential_count}")

# Step 7: Create visualizations for Recommendation 1
print("\nCreating visualizations for Recommendation 1...")

# Visualization 1: Scatter plot of facilities by CDR and volatility
plt.figure(figsize=(12, 8))
sns.scatterplot(
    data=facility_metrics,
    x='Facility_CDR',
    y='CDR_Volatility',
    hue='High_Opportunity',
    palette={True: 'red', False: 'blue'},
    alpha=0.6
)
plt.axvline(x=cdr_threshold, color='gray', linestyle='--')
plt.axhline(y=volatility_threshold, color='gray', linestyle='--')
plt.title('Facilities by Contractor Dependency Ratio (CDR) and Volatility', fontsize=16)
plt.xlabel('Contractor Dependency Ratio', fontsize=14)
plt.ylabel('CDR Volatility (Standard Deviation)', fontsize=14)
plt.annotate(f'High Opportunity\nFacilities: {high_opp_count}',
             xy=(0.85, 0.85), xycoords='axes fraction',
             bbox=dict(boxstyle="round,pad=0.3", fc="red", alpha=0.3))
plt.tight_layout()
plt.savefig('recommendation1_scatter.png', dpi=300)
plt.close()

# Visualization 2: Geographic distribution of high-opportunity facilities
plt.figure(figsize=(12, 8))
sns.barplot(
    data=geo_dist,
    x='STATE',
    y='Percentage',
    palette='viridis'
)
plt.title('Geographic Distribution of High Opportunity Facilities', fontsize=16)
plt.xlabel('State', fontsize=14)
plt.ylabel('Percentage of High Opportunity Facilities', fontsize=14)
plt.xticks(rotation=45)
for i, v in enumerate(geo_dist['Percentage']):
    plt.text(i, v + 0.5, f"{v:.1f}%", ha='center')
plt.tight_layout()
plt.savefig('recommendation1_geographic.png', dpi=300)
plt.close()

# Step 8: Create visualizations for Recommendation 2
print("\nCreating visualizations for Recommendation 2...")

# Visualization 3: Contractor utilization by staff type
contractor_by_staff = pd.DataFrame({
    'Position': ['RN', 'LPN', 'CNA'],
    'Contractor_Percentage': [
        staffing_by_type['RN_Contractor_Pct'].mean() * 100,
        staffing_by_type['LPN_Contractor_Pct'].mean() * 100,
        staffing_by_type['CNA_Contractor_Pct'].mean() * 100
    ]
})

plt.figure(figsize=(10, 6))
sns.barplot(
    data=contractor_by_staff,
    x='Position',
    y='Contractor_Percentage',
    palette='viridis'
)
plt.title('Contractor Utilization by Staff Type', fontsize=16)
plt.xlabel('Staff Position', fontsize=14)
plt.ylabel('Contractor Percentage', fontsize=14)
for i, v in enumerate(contractor_by_staff['Contractor_Percentage']):
    plt.text(i, v + 0.5, f"{v:.1f}%", ha='center')
plt.tight_layout()
plt.savefig('recommendation2_contractor_utilization.png', dpi=300)
plt.close()

# Visualization 4: Weekend vs Weekday staffing by position
weekend_weekday_data = pd.DataFrame({
    'Position': ['RN', 'LPN', 'CNA'],
    'Weekend_Ratio': [
        weekday_weekend_flat['RN_HPPD_Ratio'].mean() * 100,
        weekday_weekend_flat['LPN_HPPD_Ratio'].mean() * 100,
        weekday_weekend_flat['CNA_HPPD_Ratio'].mean() * 100
    ]
})

plt.figure(figsize=(10, 6))
sns.barplot(
    data=weekend_weekday_data,
    x='Position',
    y='Weekend_Ratio',
    palette='viridis'
)
plt.axhline(y=100, color='red', linestyle='--')
plt.title('Weekend Staffing as Percentage of Weekday Staffing', fontsize=16)
plt.xlabel('Staff Position', fontsize=14)
plt.ylabel('Weekend/Weekday Ratio (%)', fontsize=14)
for i, v in enumerate(weekend_weekday_data['Weekend_Ratio']):
    plt.text(i, v + 1, f"{v:.1f}%", ha='center')
plt.tight_layout()
plt.savefig('recommendation2_weekend_staffing.png', dpi=300)
plt.close()

# Visualization 5: Facilities with critical staffing gaps
staffing_gaps = pd.DataFrame({
    'Position': ['RN', 'LPN', 'CNA'],
    'Critical_Gap_Percentage': [
        weekday_weekend_flat['RN_Critical_Gap'].mean() * 100,
        weekday_weekend_flat['LPN_Critical_Gap'].mean() * 100,
        weekday_weekend_flat['CNA_Critical_Gap'].mean() * 100
    ]
})

plt.figure(figsize=(10, 6))
sns.barplot(
    data=staffing_gaps,
    x='Position',
    y='Critical_Gap_Percentage',
    palette='viridis'
)
plt.title('Percentage of Facilities with Critical Staffing Gaps by Position', fontsize=16)
plt.xlabel('Staff Position', fontsize=14)
plt.ylabel('Percentage of Facilities', fontsize=14)
for i, v in enumerate(staffing_gaps['Critical_Gap_Percentage']):
    plt.text(i, v + 1, f"{v:.1f}%", ha='center')
plt.tight_layout()
plt.savefig('recommendation2_critical_gaps.png', dpi=300)
plt.close()

# Generate summary statistics for report
print("\nGenerating summary statistics...")

# Overall statistics
total_facilities = len(facility_metrics)
total_contractor_hours = facility_metrics['Contractor_Hours'].sum()
total_hours = facility_metrics['Total_Hours'].sum()
overall_cdr = total_contractor_hours / total_hours

print(f"Total facilities analyzed: {total_facilities}")
print(f"Total contractor hours: {total_contractor_hours:,.0f}")
print(f"Overall contractor dependency ratio: {overall_cdr:.1%}")

# High-opportunity facilities statistics
high_opp_facilities = facility_metrics[facility_metrics['High_Opportunity']]
high_opp_contractor_hours = high_opp_facilities['Contractor_Hours'].sum()
high_opp_total_hours = high_opp_facilities['Total_Hours'].sum()
high_opp_cdr = high_opp_contractor_hours / high_opp_total_hours

print(f"High-opportunity facilities: {len(high_opp_facilities)}")
print(f"High-opportunity facilities CDR: {high_opp_cdr:.1%}")

# Staffing gap statistics
rn_gap_pct = weekday_weekend_flat['RN_Critical_Gap'].mean() * 100
lpn_gap_pct = weekday_weekend_flat['LPN_Critical_Gap'].mean() * 100
cna_gap_pct = weekday_weekend_flat['CNA_Critical_Gap'].mean() * 100

print(f"Facilities with critical RN gaps: {rn_gap_pct:.1f}%")
print(f"Facilities with critical LPN gaps: {lpn_gap_pct:.1f}%")
print(f"Facilities with critical CNA gaps: {cna_gap_pct:.1f}%")

# Weekend staffing statistics
rn_weekend_ratio = weekday_weekend_flat['RN_HPPD_Ratio'].mean() * 100
lpn_weekend_ratio = weekday_weekend_flat['LPN_HPPD_Ratio'].mean() * 100
cna_weekend_ratio = weekday_weekend_flat['CNA_HPPD_Ratio'].mean() * 100

print(f"Weekend RN staffing (% of weekday): {rn_weekend_ratio:.1f}%")
print(f"Weekend LPN staffing (% of weekday): {lpn_weekend_ratio:.1f}%")
print(f"Weekend CNA staffing (% of weekday): {cna_weekend_ratio:.1f}%")

# Contractor utilization statistics
rn_contractor_pct = staffing_by_type['RN_Contractor_Pct'].mean() * 100
lpn_contractor_pct = staffing_by_type['LPN_Contractor_Pct'].mean() * 100
cna_contractor_pct = staffing_by_type['CNA_Contractor_Pct'].mean() * 100

print(f"RN contractor utilization: {rn_contractor_pct:.1f}%")
print(f"LPN contractor utilization: {lpn_contractor_pct:.1f}%")
print(f"CNA contractor utilization: {cna_contractor_pct:.1f}%")

print("\nAnalysis complete. Visualizations saved as PNG files.")

Calculating total hours by staff type...


<ipython-input-18-e504f83fbfc2>:30: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  pbj_df['CDR'].fillna(0, inplace=True)  # Replace NaN with 0 where Total_Hrs might be 0


Merging datasets...

Analyzing contractor dependency and volatility...


<ipython-input-18-e504f83fbfc2>:59: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  facility_metrics['Facility_CDR'].fillna(0, inplace=True)
<ipython-input-18-e504f83fbfc2>:68: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)'

National average contractor dependency ratio: 6.6%
High-opportunity facilities: 341 (2.3% of total)

Analyzing staffing gaps by position type...
High-potential facilities for LPN/CNA program: 120

Creating visualizations for Recommendation 1...


<ipython-input-18-e504f83fbfc2>:233: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `x` variable to `hue` and set `legend=False` for the same effect.

  sns.barplot(



Creating visualizations for Recommendation 2...


<ipython-input-18-e504f83fbfc2>:263: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `x` variable to `hue` and set `legend=False` for the same effect.

  sns.barplot(
<ipython-input-18-e504f83fbfc2>:289: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `x` variable to `hue` and set `legend=False` for the same effect.

  sns.barplot(
<ipython-input-18-e504f83fbfc2>:316: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `x` variable to `hue` and set `legend=False` for the same effect.

  sns.barplot(



Generating summary statistics...
Total facilities analyzed: 14600
Total contractor hours: 28,904,402
Overall contractor dependency ratio: 7.0%
High-opportunity facilities: 341
High-opportunity facilities CDR: 43.8%
Facilities with critical RN gaps: 1.3%
Facilities with critical LPN gaps: 0.7%
Facilities with critical CNA gaps: 0.2%
Weekend RN staffing (% of weekday): 60.4%
Weekend LPN staffing (% of weekday): 37.0%
Weekend CNA staffing (% of weekday): 36.9%
RN contractor utilization: 7.6%
LPN contractor utilization: 9.5%
CNA contractor utilization: 7.0%

Analysis complete. Visualizations saved as PNG files.


CMS Payroll Based Journal (PBJ) Data Analysis Summary
Analysis Methodology
Data Preparation

Loaded and merged the PBJ and Provider Information datasets
Created time-based features (week numbers, weekend indicators)
Calculated key metrics:

Total hours by staff type (RN, LPN, CNA)
Contractor dependency ratio (CDR) = contractor hours / total hours
Hours per patient day (HPPD)



Opportunity Analysis

Recommendation 1: Calculated facility-level metrics to identify high contractor dependency

Measured weekly CDR volatility patterns
Identified facilities with both high CDR (>35%) and high volatility
Analyzed geographic distribution


Recommendation 2: Analyzed staffing gaps by position

Calculated staffing levels by position type
Compared weekday vs. weekend staffing patterns
Identified facilities with critical staffing gaps and low contractor utilization



Visualization Development

Created supporting visualizations including:

Facility scatter plot by CDR and volatility
Geographic distribution of high-opportunity facilities
Contractor utilization charts by staff type
Weekend vs. weekday staffing comparison
Critical staffing gap analysis



Key Findings
Overall Contractor Usage

Total contractor hours across all facilities: 28,904,402
Overall contractor dependency ratio: 7.0%

High-Opportunity Facilities (Recommendation 1)

Identified 341 facilities as high-opportunity targets
These facilities have an average CDR of 43.8% (6× the national average)
Represent clear, immediate targets for Clipboard Health's sales team

Staffing Gaps (Recommendation 2)

Critical staffing gaps exist in:

1.3% of facilities for RNs
0.7% for LPNs
0.2% for CNAs


Weekend staffing significantly lower than weekday staffing:

Weekend RN staffing: 60.4% of weekday levels
Weekend LPN staffing: 37.0% of weekday levels
Weekend CNA staffing: 36.9% of weekday levels


This reveals a major opportunity for weekend staffing solutions

Contractor Utilization by Position

RN contractor utilization: 7.6%
LPN contractor utilization: 9.5%
CNA contractor utilization: 7.0%
Despite significant weekend staffing gaps, contractor usage remains relatively low